loop: handle EAGAIN on LOOP_SET_STATUS64

poettering opened the pull request:
Since
https://github.com/torvalds/linux/commit/5db470e229e22b7eda6e23b5566e532c96fb5bc3 (i.e. kernel 5.0)
changing the .lo_offset field via LOOP_SET_STATUS64 might result in
EAGAIN. Let's handle that.

Fixes: #16858

keszybz commented:
LGTM.

keszybz commented:
bionic-ii386 timed out on slow qemu in "upstream" test, looks unrelated.

mrc0mmand commented:
Hm, something here breaks the TEST-50-DISSECT on CentOS 7 (3.10.0-1127.19.1.el7), but I can't pinpoint why. It simply times out when waiting for the loop device to appear. So far it happened twice in a row, so I don't think it's a flake.

Full journal in question: https://ci.centos.org/job/systemd-pr-build/13590/artifact//systemd-centos-ci/artifacts_nFWgaD/testsuite-logs-upstream.PmY/TEST-50-DISSECT/system.journal

Will add the second journal once the rescheduled run finishes.

keszybz commented:
If you have the journal for a passing test, please also attach for comparison.

keszybz commented:
```
[    6.836118] testsuite-50.sh[478]: + set -o pipefail
[    6.836118] testsuite-50.sh[478]: + export SYSTEMD_LOG_LEVEL=debug
[    6.836118] testsuite-50.sh[478]: + SYSTEMD_LOG_LEVEL=debug
[    6.836118] testsuite-50.sh[478]: + cd /tmp
[    6.855488] testsuite-50.sh[493]: ++ mktemp -d -t -p /tmp tmp.XXXXXX
[    6.859682] testsuite-50.sh[478]: + image_dir=/tmp/tmp.VDriq0
[    6.859682] testsuite-50.sh[478]: + '[' -z /tmp/tmp.VDriq0 ']'
[    6.859682] testsuite-50.sh[478]: + '[' '!' -d /tmp/tmp.VDriq0 ']'
[    6.859682] testsuite-50.sh[478]: + trap cleanup EXIT
[    6.859682] testsuite-50.sh[478]: + cp /usr/share/minimal.raw /usr/share/minimal.roothash /usr/share/minimal.verity /tmp/tmp.VDriq0/
[    6.906823] testsuite-50.sh[478]: + image=/tmp/tmp.VDriq0/minimal
[    6.909914] testsuite-50.sh[505]: ++ cat /tmp/tmp.VDriq0/minimal.roothash
[    6.913538] testsuite-50.sh[478]: + roothash=5a38c229b0299caf8b438b61b39100bf2098f79a200ec58545ac5f0c264c2720
[    6.913729] testsuite-50.sh[507]: + systemd-dissect --json=short /tmp/tmp.VDriq0/minimal.raw
[    6.914423] testsuite-50.sh[508]: + grep -q -F '{"rw":"ro","designator":"root","partition_uuid":null,"fstype":"squashfs","architecture":null,"verity":"external"'
[    6.939832] testsuite-50.sh[507]: Waiting for device (parent + 0 partitions) to appear...
[    6.940088] testsuite-50.sh[507]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
```
Hmmm.

keszybz commented:
On my local machine (without this PR):
```
[    5.069818] testsuite-50.sh[474]: + set -o pipefail
[    5.069818] testsuite-50.sh[474]: + export SYSTEMD_LOG_LEVEL=debug
[    5.069818] testsuite-50.sh[474]: + SYSTEMD_LOG_LEVEL=debug
[    5.069818] testsuite-50.sh[474]: + cd /tmp
[    5.070620] testsuite-50.sh[475]: ++ mktemp -d -t -p /tmp tmp.XXXXXX
[    5.085985] testsuite-50.sh[474]: + image_dir=/tmp/tmp.5Ff3p9
[    5.085985] testsuite-50.sh[474]: + '[' -z /tmp/tmp.5Ff3p9 ']'
[    5.085985] testsuite-50.sh[474]: + '[' '!' -d /tmp/tmp.5Ff3p9 ']'
[    5.085985] testsuite-50.sh[474]: + trap cleanup EXIT
[    5.085985] testsuite-50.sh[474]: + cp /usr/share/minimal.raw /usr/share/minimal.roothash /usr/share/minimal.verity /tmp/tmp.5Ff3p9/
[    5.103978] testsuite-50.sh[474]: + image=/tmp/tmp.5Ff3p9/minimal
[    5.104177] testsuite-50.sh[479]: ++ cat /tmp/tmp.5Ff3p9/minimal.roothash
[    5.108429] testsuite-50.sh[474]: + roothash=ad24f9a91f30dea9d4f375c67f8150b41ef41dd885d4c84b4362d436b9f84a7a
[    5.108872] testsuite-50.sh[481]: + grep -q -F '{"rw":"ro","designator":"root","partition_uuid":null,"fstype":"squashfs","architecture":null,"verity":"external"'
[    5.109415] testsuite-50.sh[480]: + systemd-dissect --json=short /tmp/tmp.5Ff3p9/minimal.raw
[    5.141133] testsuite-50.sh[480]: Waiting for device (parent + 0 partitions) to appear...
[    5.196437] testsuite-50.sh[480]: Successfully forked off '(sd-dissect)' as PID 487.
[    5.196831] testsuite-50.sh[487]: Mounting squashfs on /tmp/dissect-gQrWXv (MS_RDONLY|MS_NODEV "")...
[    5.228944] testsuite-50.sh[487]: Failed to read /etc/hostname file of image, ignoring: No such file or directory
[    5.228944] testsuite-50.sh[487]: Failed to read /etc/machine-id file of image, ignoring: No such file or directory
[    5.229176] testsuite-50.sh[487]: Failed to read /etc/machine-info file of image, ignoring: No such file or directory
[    5.229382] testsuite-50.sh[480]: Failed to read /etc/hostname: No such file or directory
[    5.229382] testsuite-50.sh[480]: /etc/machine-id file is empty.
[    5.230589] testsuite-50.sh[480]: (sd-dissect) succeeded.
[    5.246167] testsuite-50.sh[491]: + systemd-dissect /tmp/tmp.5Ff3p9/minimal.raw
[    5.246630] testsuite-50.sh[492]: + grep -q -F MARKER=1
[    5.261228] testsuite-50.sh[491]: Waiting for device (parent + 0 partitions) to appear...
[    5.261927] testsuite-50.sh[491]: Successfully forked off '(sd-dissect)' as PID 494.
[    5.262121] testsuite-50.sh[494]: Mounting squashfs on /tmp/dissect-Os16nl (MS_RDONLY|MS_NODEV "")...
[    5.262710] testsuite-50.sh[494]: Failed to read /etc/hostname file of image, ignoring: No such file or directory
[    5.262710] testsuite-50.sh[494]: Failed to read /etc/machine-id file of image, ignoring: No such file or directory
[    5.262929] testsuite-50.sh[494]: Failed to read /etc/machine-info file of image, ignoring: No such file or directory
[    5.263032] testsuite-50.sh[491]: Failed to read /etc/hostname: No such file or directory
[    5.263162] testsuite-50.sh[491]: /etc/machine-id file is empty.
[    5.263925] testsuite-50.sh[491]: (sd-dissect) succeeded.
[    5.277395] testsuite-50.sh[495]: + systemd-dissect /tmp/tmp.5Ff3p9/minimal.raw
[    5.278029] testsuite-50.sh[496]: + grep -q -F -f /usr/lib/os-release
[    5.279434] testsuite-50.sh[496]: grep: /usr/lib/os-release: No such file or directory
[    5.290903] testsuite-50.sh[495]: Waiting for device (parent + 0 partitions) to appear...
[    5.292982] testsuite-50.sh[495]: Successfully forked off '(sd-dissect)' as PID 498.
[    5.293141] testsuite-50.sh[498]: Mounting squashfs on /tmp/dissect-6LEe4B (MS_RDONLY|MS_NODEV "")...
[    5.293717] testsuite-50.sh[498]: Failed to read /etc/hostname file of image, ignoring: No such file or directory
[    5.293717] testsuite-50.sh[498]: Failed to read /etc/machine-id file of image, ignoring: No such file or directory
[    5.293886] testsuite-50.sh[498]: Failed to read /etc/machine-info file of image, ignoring: No such file or directory
[    5.293995] testsuite-50.sh[495]: Failed to read /etc/hostname: No such file or directory
[    5.294118] testsuite-50.sh[495]: /etc/machine-id file is empty.
[    5.295126] testsuite-50.sh[495]: (sd-dissect) succeeded.
[    5.295323] testsuite-50.sh[495]: Failed to dump table: Broken pipe
[    5.306226] testsuite-50.sh[474]: + cleanup
[    5.306226] testsuite-50.sh[474]: + '[' -z /tmp/tmp.5Ff3p9 ']'
[    5.306475] testsuite-50.sh[474]: + rm -rf /tmp/tmp.5Ff3p9
```
(also failed, but much later)

mrc0mmand commented:
Manually procured passing test journal from the latest master on the *same* machine: [systemd.journal.tar.gz](https://github.com/systemd/systemd/files/5135651/systemd.journal.tar.gz)

keszybz commented:
The failure on my machine had a different cause. I filed https://github.com/systemd/systemd/pull/16872 to deal with that.

OK, so it seems that with this PR systemd-dissect "hangs". It might be caused by in issue in it, or it might be caused by a failure to unlock somewhere else. Either way we should figure this out before merging.

poettering commented:
So, my educated guess is that we take the BSD lock, then release it and this means udev won't probe our loopback device if it tries to do so in this time window. Normally, it would retry it, because it watches close() events, and reprobes on each. But if we keep the fd open we just unlocked and then wait for the loopback device to be fully probed we can wait forever, since no close() is done and ths udev will not reprobe it. If you follow what I blubbering here...

I force pushed a new version that should fix that: it issues the BSD lock on a separately opened device fd. When we release the lock we also close that separate fd, which will then retrigger and we can wait for it properly. Turns out doing it like that is actually also prettier, than my earlier PR, since we can use a simple `_cleanup_close_` for automatically releasing the lock + closing the fd.

yuwata commented on src/shared/loop-util.c:
```diff
@@ -84,15 +85,55 @@ static int loop_configure(int fd, const struct loop_config *c) {
+         * EAGAIN and retry. The former should be an efficient mechanism to avoid we have to wait 50ms
+         * needlessly if we are just racing against udev. The latter is protection against all other cases,
+         * i.e. peers that do not take the BSD lock. */
+        lock_fd =
+                fd_reopen(fd, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY); /* We take the BSD lock on a second,
+                                                                        * separately opened fd for the
+                                                                        * device. udev after all watches for
+                                                                        * close() events on block devices to
+                                                                        * reprobe them, hence by having a
+                                                                        * separate fd we will later close()
+                                                                        * we can ensure we trigger udev after
+                                                                        * everything is done. If we'd lock
+                                                                        * our own fd instead and keep it open
+                                                                        * for a long time udev would possibly
+                                                                        * never run on it again, even though
+                                                                        * the fd is unlocked, simply because
+                                                                        * we never close() it. It also has
+                                                                        * the nice benefit we can use the
+                                                                        * _cleanup_close_ logic to
+                                                                        * automatically release the lock,
+                                                                        * after we are done. */
```
I do not like this indentation...

yuwata reviewed (approved):
LGTM.

mrc0mmand commented:
Unfortunately, TEST-50-DISSECT still hangs when waiting for the loop device.

poettering commented:
this weird, i have no idea!

mrc0mmand commented:
@poettering I went a bit deeper in hopes to get this one finally sorted out and apart from it happening intermittently (like in every second/third attempt), I managed to grab an strace during one of the hangs:

```
[    7.564350] testsuite-50.sh[568]: fstat(4, {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
[    7.564350] testsuite-50.sh[568]: close(3)                                = 0
[    7.564350] testsuite-50.sh[568]: openat(4, "dev", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 3
[    7.564350] testsuite-50.sh[568]: fstat(3, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
[    7.564350] testsuite-50.sh[568]: close(4)                                = 0
[    7.564350] testsuite-50.sh[568]: openat(3, "block", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 4
[    7.564350] testsuite-50.sh[568]: fstat(4, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
[    7.564350] testsuite-50.sh[568]: close(3)                                = 0
[    7.565307] testsuite-50.sh[568]: openat(4, "7:0", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 3
[    7.565307] testsuite-50.sh[568]: fstat(3, {st_mode=S_IFLNK|0777, st_size=0, ...}) = 0
[    7.565307] testsuite-50.sh[568]: readlinkat(4, "7:0", "../../devices/virtual/block/loop0", 4096) = 33
[    7.565307] testsuite-50.sh[568]: close(3)                                = 0
[    7.565307] testsuite-50.sh[568]: openat(4, "..", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 3
[    7.565307] testsuite-50.sh[568]: close(4)                                = 0
[    7.565307] testsuite-50.sh[568]: openat(3, "..", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 4
[    7.565307] testsuite-50.sh[568]: close(3)                                = 0
[    7.565307] testsuite-50.sh[568]: openat(4, "devices", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 3
[    7.565307] testsuite-50.sh[568]: fstat(3, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
[    7.565307] testsuite-50.sh[568]: close(4)                                = 0
[    7.565307] testsuite-50.sh[568]: openat(3, "virtual", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 4
[    7.565307] testsuite-50.sh[568]: fstat(4, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
[    7.565307] testsuite-50.sh[568]: close(3)                                = 0
[    7.565307] testsuite-50.sh[568]: openat(4, "block", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 3
[    7.565307] testsuite-50.sh[568]: fstat(3, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
[    7.565307] testsuite-50.sh[568]: close(4)                                = 0
[    7.565307] testsuite-50.sh[568]: openat(3, "loop0", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 4
[    7.565307] testsuite-50.sh[568]: fstat(4, {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
[    7.565307] testsuite-50.sh[568]: close(3)                                = 0
[    7.565307] testsuite-50.sh[568]: close(4)                                = 0
[    7.565307] testsuite-50.sh[568]: access("/sys/devices/virtual/block/loop0/uevent", F_OK) = 0
[    7.565307] testsuite-50.sh[568]: gettid()                                = 571
[    7.565307] testsuite-50.sh[568]: getrandom(0x7f0ae7dd61a0, 16, GRND_NONBLOCK|0x4) = -1 EINVAL (Invalid argument)
[    7.565307] testsuite-50.sh[568]: getrandom("\xda\x5b\x98\xa7\x01\xc2\x68\x44\x30\x66\xb4\x8b\xe3\x0a\x26\x08", 16, GRND_NONBLOCK) = 16
[    7.567231] testsuite-50.sh[568]: writev(2, [{iov_base="Waiting for device (parent + 0 partitions) to appear...", iov_len=55}, {iov_base="\n", iov_len=1}], 2
[    7.567231] testsuite-50.sh[571]: Waiting for device (parent + 0 partitions) to appear...
[    7.568283] testsuite-50.sh[568]: ) = 56
[    7.568283] testsuite-50.sh[568]: readlinkat(AT_FDCWD, "/sys/devices/virtual/block/loop0/subsystem", "../../../../class/block", 4096) = 23
[    7.568283] testsuite-50.sh[568]: getrandom(0x7f0ae7dd6180, 16, GRND_NONBLOCK|0x4) = -1 EINVAL (Invalid argument)
[    7.568283] testsuite-50.sh[568]: getrandom("\x22\x2e\x7b\x67\xde\xb5\xa4\xe0\x1f\xc9\x0b\xa1\xf6\x4f\xad\x8c", 16, GRND_NONBLOCK) = 16
[    7.568283] testsuite-50.sh[568]: open("/sys/devices/virtual/block/loop0/uevent", O_RDONLY|O_CLOEXEC) = 3
[    7.568283] testsuite-50.sh[568]: fstat(3, {st_mode=S_IFREG|0644, st_size=4096, ...}) = 0
[    7.568283] testsuite-50.sh[568]: fstat(3, {st_mode=S_IFREG|0644, st_size=4096, ...}) = 0
[    7.568283] testsuite-50.sh[568]: mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0ae7ff5000
[    7.568283] testsuite-50.sh[568]: read(3, "MAJOR=7\nMINOR=0\nDEVNAME=loop0\nDEVTYPE=disk\n", 4096) = 43
[    7.568283] testsuite-50.sh[568]: read(3, "", 4096)                       = 0
[    7.568283] testsuite-50.sh[568]: close(3)                                = 0
[    7.568283] testsuite-50.sh[568]: munmap(0x7f0ae7ff5000, 4096)            = 0
[    7.568283] testsuite-50.sh[568]: getrandom(0x7f0ae7dd6180, 16, GRND_NONBLOCK|0x4) = -1 EINVAL (Invalid argument)
[    7.568283] testsuite-50.sh[568]: getrandom("\xce\xbe\xb9\x7b\xbe\x5f\xe7\xc2\x98\x6a\xb0\xfc\xaf\x9e\xe2\xf7", 16, GRND_NONBLOCK) = 16
[    7.568283] testsuite-50.sh[568]: open("/run/udev/data/b7:0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[    7.568283] testsuite-50.sh[568]: epoll_create1(EPOLL_CLOEXEC)            = 3
[    7.568283] testsuite-50.sh[568]: access("/run/udev/control", F_OK)       = 0
[    7.568283] testsuite-50.sh[568]: socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC|SOCK_NONBLOCK, NETLINK_KOBJECT_UEVENT) = 4
[    7.568283] testsuite-50.sh[568]: ioctl(4, SIOCGSKNS, 0)                  = -1 EOPNOTSUPP (Operation not supported)
[    7.569767] testsuite-50.sh[568]: writev(2, [{iov_base="sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported", iov_len=143}, {iov_base="\n", iov_len=1}], 2
[    7.569767] testsuite-50.sh[571]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
[    7.569962] testsuite-50.sh[568]: ) = 144
[    7.569962] testsuite-50.sh[568]: setsockopt(4, SOL_SOCKET, SO_ATTACH_FILTER, {len=8, filter=0x7ffe82f81f60}, 16) = 0
[    7.569962] testsuite-50.sh[568]: setsockopt(4, SOL_SOCKET, SO_PASSCRED, [1], 4) = 0
[    7.569962] testsuite-50.sh[568]: bind(4, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=0x000002}, 12) = 0
[    7.569962] testsuite-50.sh[568]: getsockname(4, {sa_family=AF_NETLINK, nl_pid=571, nl_groups=0x000002}, [12]) = 0
[    7.569962] testsuite-50.sh[568]: epoll_ctl(3, EPOLL_CTL_ADD, 4, {EPOLLIN, {u32=8066656, u64=8066656}}) = 0
[    7.569962] testsuite-50.sh[568]: open("/run/udev/data/b7:0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

Full journal (strace included): [system.journal.tar.gz](https://github.com/systemd/systemd/files/5150209/system.journal.tar.gz)

poettering commented:
I might have figured it out. udev watches the devices with IN_CLOSE_WRITE, hence we have to open the block device in write mode, O_RDONLY does not suffice when we are interested in udev reprobing on close().

I pushed a new version with that fixed (i.e. O_RDONLY → O_RDWR when opening the block device the second time for the BSD lock). PTAL.

mrc0mmand commented:
Unfortunately, it still hangs on CentOS 7:

```
[    7.390615] testsuite-50.sh[529]: + systemd-dissect /tmp/tmp.pjnsrm/minimal.raw --root-hash=5eda439297dedb96cc70f571eb4b4c28b5d62269f0d08a849a39d61a24a04400 --verity-data=/tmp/tmp.pjnsrm/minimal.fooverity
[    7.391888] systemd[1]: sys-devices-virtual-block-loop0.device: Installed new job sys-devices-virtual-block-loop0.device/nop as 281
[    7.392143] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=106 reply_cookie=0 signature>
[    7.392305] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=107 reply_cookie=0 signature>
[    7.392455] systemd[1]: dev-loop0.device: Changed plugged -> dead
[    7.392627] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    7.392792] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    7.392941] systemd[1]: sys-devices-virtual-block-loop0.device: Changed plugged -> dead
[    7.393136] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    7.393290] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    7.393437] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=112 reply_cookie=0 signature>
[    7.393630] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=113 reply_cookie=0 signature>
[    7.393778] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobNew cookie=114 reply_cookie=0 signature=uos error-name=n/a error-message=n/a
[    7.393922] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobNew cookie=115 reply_cookie=0 signature=uos error-name=n/a error-message=n/a
[    7.394137] systemd[1]: dev-loop0.device: Job 280 dev-loop0.device/nop finished, result=done
[    7.394298] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=116 reply_cookie=0 signature=uoss error-name=n/a error-messa>
[    7.394458] systemd[1]: sys-devices-virtual-block-loop0.device: Job 281 sys-devices-virtual-block-loop0.device/nop finished, result=done
[    7.394632] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=117 reply_cookie=0 signature=uoss error-name=n/a error-messa>
[    7.394781] systemd[1]: sys-devices-virtual-block-loop0.device: Collecting.
[    7.394929] systemd[1]: dev-loop0.device: Collecting.
[    7.395135] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=118 reply_cookie=0 signature>
[    7.395290] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=119 reply_cookie=0 signature>
[    7.395451] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=120 reply_cookie=0 signature=so error-name=n/a error-messag>
[    7.395617] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    7.395765] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dloop0_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=>
[    7.395910] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=123 reply_cookie=0 signature=so error-name=n/a error-messag>
[    7.403059] testsuite-50.sh[529]: Waiting for device (parent + 0 partitions) to appear...
[    7.403282] testsuite-50.sh[529]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
```

Is there anything I could gather to make the debugging easier?

poettering commented:
hmm, any chance you can get udev debug logs when this happens on that machine?

i.e. for some reason udev doesn't reprobe the device when we issue our close(), and the question is why, i.e. whether it doesn't see it or what else happened there...

mrc0mmand commented:
@poettering sure, here's a snippet from around the hang, I'll upload the whole journal shortly:

```
[    7.648519] testsuite-50.sh[554]: + systemd-dissect /tmp/tmp.INvTyK/minimal.raw
[    7.649249] testsuite-50.sh[555]: + grep -q -F -f /etc/os-release
[    7.649851] systemd-udevd[473]: loop0: Handling device node '/dev/loop0', devnum=b7:0
[    7.650129] systemd-udevd[473]: loop0: Preserve already existing symlink '/dev/block/7:0' to '../loop0'
[    7.650228] systemd-udevd[473]: loop0: sd-device: Created db file '/run/udev/data/b7:0' for '/devices/virtual/block/loop0'
[    7.650339] systemd-udevd[473]: loop0: Adding watch on '/dev/loop0'
[    7.650430] systemd-udevd[473]: loop0: Failed to add device '/dev/loop0' to watch: No such file or directory
[    7.650520] systemd-udevd[473]: loop0: sd-device: Created db file '/run/udev/data/b7:0' for '/devices/virtual/block/loop0'
[    7.650608] systemd-udevd[473]: loop0: Device (SEQNUM=1810, ACTION=change) processed
[    7.650698] systemd-udevd[473]: loop0: sd-device-monitor: Passed to netlink monitor
[    7.650899] systemd-udevd[467]: loop0: Device (SEQNUM=1812, ACTION=remove) ready for processing
[    7.651027] systemd-udevd[467]: loop0: sd-device-monitor: Passed 192 byte to netlink monitor
[    7.651120] systemd[1]: Added job dev-loop0.device/nop to transaction.
[    7.651206] systemd[1]: dev-loop0.device: Installed new job dev-loop0.device/nop as 274
[    7.651302] systemd[1]: Added job sys-devices-virtual-block-loop0.device/nop to transaction.
[    7.651386] systemd[1]: sys-devices-virtual-block-loop0.device: Installed new job sys-devices-virtual-block-loop0.device/nop as 275
[    7.651474] systemd-udevd[473]: loop0: Processing device (SEQNUM=1812, ACTION=remove)
[    7.651661] systemd-udevd[473]: loop0: Device (SEQNUM=1812, ACTION=remove) processed
[    7.651749] systemd-udevd[473]: loop0: sd-device-monitor: Passed to netlink monitor
[    7.651923] systemd[1]: dev-loop0.device: Changed plugged -> dead
[    7.652204] systemd[1]: sys-devices-virtual-block-loop0.device: Changed plugged -> dead
[    7.653161] systemd[1]: dev-loop0.device: Job 274 dev-loop0.device/nop finished, result=done
[    7.653357] systemd[1]: sys-devices-virtual-block-loop0.device: Job 275 sys-devices-virtual-block-loop0.device/nop finished, result=done
[    7.653572] systemd[1]: sys-devices-virtual-block-loop0.device: Collecting.
[    7.653670] systemd[1]: dev-loop0.device: Collecting.
[    7.659392] systemd-udevd[467]: 7:0: Device (SEQNUM=1813, ACTION=add) is queued
[    7.659612] systemd-udevd[467]: 7:0: Device (SEQNUM=1813, ACTION=add) ready for processing
[    7.659767] systemd-udevd[467]: 7:0: sd-device-monitor: Passed 135 byte to netlink monitor
[    7.659932] systemd-udevd[467]: loop0: Device (SEQNUM=1814, ACTION=add) is queued
[    7.660186] systemd-udevd[467]: loop0: Device (SEQNUM=1814, ACTION=add) ready for processing
[    7.660407] systemd-udevd[467]: loop0: sd-device-monitor: Passed 189 byte to netlink monitor
[    7.660594] systemd-udevd[473]: 7:0: Processing device (SEQNUM=1813, ACTION=add)
[    7.660785] systemd-udevd[467]: loop0: Device (SEQNUM=1815, ACTION=change) is queued
[    7.660982] systemd-udevd[467]: loop0: SEQNUM=1815 blocked by SEQNUM=1814
[    7.661234] systemd-udevd[474]: loop0: Processing device (SEQNUM=1814, ACTION=add)
[    7.661451] systemd-udevd[474]: loop0: Failed to flock(/dev/loop0): Resource temporarily unavailable
[    7.661651] systemd-udevd[474]: loop0: Device currently locked, not processing.
[    7.661855] systemd-udevd[467]: loop0: Device (SEQNUM=1815, ACTION=change) ready for processing
[    7.662273] systemd-udevd[467]: loop0: sd-device-monitor: Passed 192 byte to netlink monitor
[    7.662456] testsuite-50.sh[554]: Waiting for device (parent + 0 partitions) to appear...
[    7.662456] testsuite-50.sh[554]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
[    7.662681] systemd-udevd[474]: loop0: Processing device (SEQNUM=1815, ACTION=change)
[    7.662832] systemd-udevd[474]: loop0: Failed to flock(/dev/loop0): Resource temporarily unavailable
[    7.662992] systemd-udevd[474]: loop0: Device currently locked, not processing.
[    7.663201] systemd-udevd[473]: 7:0: Device (SEQNUM=1813, ACTION=add) processed
[    7.663368] systemd-udevd[473]: 7:0: sd-device-monitor: Passed to netlink monitor
```

mrc0mmand commented:
I also wonder if this issue (or the patch) is somehow related to https://github.com/systemd/systemd/issues/16806 or if it's a completely different issue, /cc @bluca.

poettering commented:
So I think I tracked it down now: if a device is locked, udev won't process udev rules. The idea is to watch for IN_CLOSE_WRITE inotify events in that case, so that udev notices when the device is not used anymore and then reprobes the device. however this inotify watching is turned on by udev rules. So if we never run the rules how can this option ever be turned on?

This wasn't noticed previously I guess because we never systematically made a device pop up and immediately locked it. Because this is a race: if udev was quick enough to run its rules at least once before the device was locked the watch would be turned on, and then stayed on for good until the device went away again.

I now added a fix for this: first of all the watches are now removed again when the rules say so. Thus on each device event you can turn on/off the watching individually and it will matter until the next event is seen. Secondly, whenever we encounter the device locked we'll now install the watch immediately, so that we definitely notice when the device is closed again.

This means inotify watching is now enabled by two separate conditions: by explicit rules *or* because the device was locked when we tried to run rules for it.

Anyway, let's see if this finally fixes things.

poettering commented:
hmm, archlinux CI failed, and not for unrelated reasons. bummer.

bluca commented:
> hmm, archlinux CI failed, and not for unrelated reasons. bummer.

And of course once debug logs are enabled it doesn't fail anymore. The machine-gods are unjust...

mrc0mmand commented:
Relevant section from the last CentOS 7 TEST-50-DISSECT fail:

```
[    9.188858] systemd[880]: Mounting squashfs on /run/systemd/unit-root (MS_RDONLY|MS_NODEV "")...
[    9.190209] systemd[880]: Applying namespace mount on /run/systemd/unit-root/run
[    9.190547] systemd[880]: Applying namespace mount on /run/systemd/unit-root/run/credentials
[    9.190786] systemd[880]: Applying namespace mount on /run/systemd/unit-root/run/img1
[    9.191213] systemd-udevd[450]: XXX: lock block dev failed: Resource temporarily unavailable
[    9.334310] kernel:  loop2: p1 p2
[    9.192124] systemd-udevd[450]: XXX: lock block dev failed: Resource temporarily unavailable
[    9.193848] systemd[880]: Waiting for device (parent + 2 partitions) to appear...
[    9.200532] systemd[880]: Mounting squashfs on /run/systemd/unit-root/run/img1 (MS_RDONLY|MS_NODEV "")...
[    9.203484] systemd[880]: Applying namespace mount on /run/systemd/unit-root/run/img2
[    9.207137] systemd[880]: Waiting for device (parent + 0 partitions) to appear...
[    9.208987] systemd[880]: Device a4009773230bb7b5bb2c5d5648f443e55ea3fb98f0727c377638e8d8b7e685f0-verity already exists.
[    9.209934] systemd[880]: Error opening verity device, crypt_volume_key_get failed: Invalid argument
[    9.232059] systemd[880]: Mounting squashfs on /run/systemd/unit-root/run/img2 (MS_RDONLY|MS_NODEV "")...
[    9.474443] systemd[1]: run-u16.service: Succeeded.
...
[    9.488832] testsuite-50.sh[899]: + systemd-run -t --property TemporaryFileSystem=/run --property RootImage=/tmp/tmp.sDuwFy/minimal.gpt --property RootHash=a4009773230bb7b5bb2c5d5648f443e55ea3fb98f0727c377638e8d8b7e685f0 --property 'MountImages=/tmp/tmp.sDuwFy/minimal.gpt:/run/img1 /tmp/tmp.sDuwFy/minimal.raw:/run>
[    9.489857] testsuite-50.sh[900]: + grep -q -F MARKER=1
...
[    9.499127] testsuite-50.sh[899]: Press ^] three times within 1s to disconnect TTY.
[    9.499127] testsuite-50.sh[899]: Assertion 's' failed at src/libsystemd/sd-event/sd-event.c:2100, function sd_event_source_set_priority(). Ignoring.
...
[    9.499127] testsuite-50.sh[899]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=3 reply_cookie=0 signature=s error-name=n/a error-message=n/a
[    9.505178] systemd[1]: Started run-u17.service.
[    9.649186] kernel:  loop0: p1 p2
[    9.507829] systemd[901]: Waiting for device (parent + 2 partitions) to appear...
...
[    9.510378] systemd[901]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
...
[    9.533938] systemd[901]: Mounting squashfs on /run/systemd/unit-root (MS_RDONLY|MS_NODEV "")...
[    9.535313] systemd[901]: Applying namespace mount on /run/systemd/unit-root/run
[    9.535513] systemd[901]: Applying namespace mount on /run/systemd/unit-root/run/credentials
[    9.535743] systemd[901]: Applying namespace mount on /run/systemd/unit-root/run/img1
[    9.678794] kernel:  loop1: p1 p2
[    9.536595] systemd[901]: Waiting for device (parent + 2 partitions) to appear...
[    9.537451] systemd[901]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
[    9.680652] kernel:  loop1: p1 p2
[    9.545599] systemd-udevd[450]: XXX: exec rules failed: File exists
[    9.545668] systemd-udevd[450]: loop1p1: Failed to process device, ignoring: File exists
[    9.545970] systemd-udevd[443]: XXX: exec rules failed: File exists
[    9.546087] systemd[901]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Operation not supported
[    9.546151] systemd-udevd[443]: loop1p2: Failed to process device, ignoring: File exists
[    9.546538] systemd[901]: Failed to mount image: No such device or address
[    9.606985] systemd[901]: run-u17.service: Failed to set up mount namespacing: /run/systemd/unit-root/run/img1: No such device or address
[    9.607172] systemd[901]: run-u17.service: Failed at step NAMESPACE spawning /usr/bin/cat: No such device or address
[    9.614480] systemd[1]: run-u17.service: Main process exited, code=exited, status=226/NAMESPACE
```

Full journal: https://ci.centos.org/job/systemd-pr-build/13797/artifact//systemd-centos-ci/artifacts_1twczu/testsuite-logs-upstream.Byn/TEST-50-DISSECT/system.journal

mrc0mmand commented:
@poettering opened https://github.com/systemd/systemd-centos-ci/pull/303 which runs TEST-50-DISSECT 100x on CentOS 7 to make the debugging faster (and hopefully easier).

poettering commented:
ah, interesting! thanks!

mrc0mmand commented:
@poettering https://jenkins-systemd.apps.ocp.ci.centos.org/job/ci-build/31/artifact//artifacts_all/artifacts_qdvI7t/testsuite-logs-upstream.yM9/TEST-50-DISSECT/system.journal

mrc0mmand commented:
Also, to make it even easier, you can comment *retest this please* on https://github.com/systemd/systemd-centos-ci/pull/303 to make it pull the new changes from this PR and run the loop again.

poettering commented:
> Also, to make it even easier, you can comment _retest this please_ on [systemd/systemd-centos-ci#303](https://github.com/systemd/systemd-centos-ci/pull/303) to make it pull the new changes from this PR and run the loop again.

Oh, super nice thanks!

poettering commented:
So this is ready for review, btw, and I think we should merge soon. The Ubuntu CI doesn't pass as of now, but that's because it uses an old udev, which doesn't carry #17211 (which are also included here), which means some udev notifications are lost.

I think we should get this stuff merged, I am pretty sure it gets us at least very close to fixing everything in loopback setup — but I think there's a good chnce it actually fixes things comprehensively.

The centos CI is down currently, which is a pity, but i hope the other CIs succeeding counts for at least something...

Anyway, ptal!

mrc0mmand commented:
> The centos CI is down currently, which is a pity, but i hope the other CIs succeeding counts for at least something...

CentOS CI is up again and I manually redelivered all failed webhook events, so the results should, hopefully, find their way here.

bluca reviewed (approved):
Diff looks good to me. Great to also have a test case. Hopefully it will help with the loop block dev errors we occasionally see in the dissect system test.

bluca commented:
Just a couple of questions for my own use case's benefit: I care a lot about the "happy path" at boot, with lots of single-fs images. As far as I can see, there shouldn't be any extra probing/waiting after these changes, as they mostly apply to partitioned devices. Is that right?
Also, would having lots of pre-allocated loop devices (CONFIG_BLK_DEV_LOOP_MIN_COUNT) help with avoiding the attach-empty-file-to-clean fallback? Or would it make it worse?

bluca commented:
> > Just a couple of questions for my own use case's benefit: I care a lot about the "happy path" at boot, with lots of single-fs images. As far as I can see, there shouldn't be any extra probing/waiting after these changes, as they mostly apply to partitioned devices. Is that right?
> 
> This makes no real changes what we wait for, just corrects how we do it. And while that might have worked incorrectly before for the partition sub block devices it always worked correctly for the main block device, and thus performance shouldn't have changed.
> 
> > Also, would having lots of pre-allocated loop devices (CONFIG_BLK_DEV_LOOP_MIN_COUNT) help with avoiding the attach-empty-file-to-clean fallback? Or would it make it worse?
> 
> I doubt it would change much. After all when the issue happens the device is detached and sticks around. In the bug condition it sticks around because of the kernel won#t collect it as long as the partition block devices are still around. If you pre-allocate devices, then they will also stay around otherwise, regardless of the bug or not...

Ok, makes sense. Thanks!

keszybz commented on src/shared/dissect-image.c:
```diff
@@ -368,6 +368,86 @@ static void check_partition_flags(
+                if (r < 0)
+                        return r;
+        }
+}
```
I guess it's OK to merge a work-around like this, but we should plan to remove it at some point. Let's say systemd-250.

keszybz commented on src/shared/loop-util.c:
```diff
@@ -414,6 +418,11 @@ int loop_device_make(
+
+                /* Wait some random time, to make collision less likely. Let's pick a random time in the
+                 * range 0ms…250ms, linearly scaled by the number of failed attempts. */
+                (void) usleep(random_u64() % (UINT64_C(10) * USEC_PER_MSEC +
```
USEC_PER_MSEC already has the necessary type, so the UINT64_C() cast is a noop. (Though it doesn't hurt.)

keszybz reviewed (approved):
I really don't like the complexity of this... I *think* the code is OK, but it just seems we're piling on workarounds instead of fixing the underlying APIs. Can we work with the kernel people to provide a sane loopback device API?

keszybz commented:
Hmm, both bionic-amd64 and bionic-i386 timed out in the new test. I'll restart them.

keszybz commented:
bionic-amd64 timed out in the new test, and bionic-i386 failed "boot-smoke".
I would prefer to skip the new test in ubuntuci. Even if the test only fails ~20% of times, it's going to be a constant source of noise.

poettering commented:
So, I added code to skip the test on Ubuntu CI for now, given the woes. And see that it passed all green now! Yay!

Given the positive reviews so far, let's merge!

yuwata commented on meson.build:
```diff
@@ -3338,6 +3338,7 @@ foreach tuple : tests
         type = tuple.length() >= 5 ? tuple[4] : ''
         defs = tuple.length() >= 6 ? tuple[5] : []
         incs = tuple.length() >= 7 ? tuple[6] : includes
+        parallel = tuple.length() >= 8 ? tuple[7] : true
```
Unused??
